File: 496-next-greater-element-i/496-next-greater-element-i.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, nums2) {
  const map = {}
  nums1.forEach((v, i) => map[v] = i)
  const ans = Array(nums1.length).fill(-1)
  const stack = []
  for (const v of nums2) {
    while (stack.length > 0 && v > stack.at(-1)) {
      if (map[stack.at(-1)] != undefined) {
        ans[map[stack.at(-1)]] = v
      }
      stack.pop()
    }
    stack.push(v)
  }
  return ans
};

File: 496-next-greater-element-i/496-next-greater-element-i.go
func nextGreaterElement(nums1 []int, nums2 []int) []int {
  ans := make([]int, len(nums1))
  l := len(nums2)
  for j, v := range nums1 {
    i := 0
    for ; nums2[i] != v; i++ {}
    for ; i < l && nums2[i] <= v; i++ {}
    if i >= l {
      ans[j] = -1
    } else {
      ans[j] = nums2[i]
    }   
  }
  return ans
}